File: daily_note_cleanup.py
#!/usr/bin/env python

# This looks at all Daily Notes and compares their contents to the template file. If the files are exactly the same as the template, they are moved to the trash.
# After trashing empty Daily Notes, we check the creation dates and move any Daily Notes older than 7 days to the Archive folder.

import os
import time

notes_path = '/Users/justinlamb/Library/Mobile Documents/iCloud~md~obsidian/Documents/Obsidian/Daily Notes'
trash_path = '/Users/justinlamb/.Trash/'
archive_path = '/Users/justinlamb/Library/Mobile Documents/iCloud~md~obsidian/Documents/Obsidian/Daily Notes/Archive'
files = os.listdir(notes_path)
template = open(os.path.join(notes_path, './Archive/DailyNoteTemplate.md'))

template_lines = template.readlines()[1:]

week = 604800
now = time.time()
cutoff = now - week 

for file in files:
    if os.path.isfile(os.path.join(notes_path, file)):
        if file != '.DS_Store':
            creation = os.path.getctime(os.path.join(notes_path, file)) # Find creation date.
            f = open(os.path.join(notes_path, file),'r')
            f_lines = f.readlines()[1:]
            if f_lines == template_lines: # Compare to template and trash empty notes.
                os.rename(os.path.join(notes_path, file), os.path.join(trash_path, file))
            elif creation < cutoff: # Archive files older than 7 days.
                os.rename(os.path.join(notes_path, file), os.path.join(archive_path, file))
            f.close()

File: daily_note_cleanup.rb
# frozen_string_literal: true

# daily_note_cleanup.py rewritten in Ruby.
# Allows flexibility of template lines, ie. date line dynamically extracted rather than removing line 1
# Moves notes which match template array exactly to trash
# Moves notes with birthtime of greater than 7 days to Archive folder

require 'fileutils'

OBSIDIAN_PATH = '/Users/justinlamb/Library/Mobile Documents/iCloud~md~obsidian/Documents/Obsidian/'
TEMPLATE_PATH = "#{OBSIDIAN_PATH}Templates/DailyNoteTemplate.md"
ARCHIVE_PATH = "#{OBSIDIAN_PATH}Daily Notes/Archive"
TRASH_PATH = '/Users/justinlamb/.Trash/'
IGNORE_LINE = ['{{date:dddd, MMMM Do YYYY}}', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday',
               'Sunday'].freeze
ARCHIVE_TIME = (24 * 3600) * 7
trash_count = 0
archive_count = 0

template = []
File.foreach(TEMPLATE_PATH) { |line| template << line unless line.match? Regexp.union(IGNORE_LINE) }

notes = Dir["#{OBSIDIAN_PATH}Daily Notes/*.md"].select

puts "📝 Checking Daily Notes"

notes.each do |file|
  note = []
  year = File.basename(file)[0..3]
  dir = "#{ARCHIVE_PATH}/#{year}"
  Dir.mkdir(dir) unless Dir.exist?(dir)
  File.foreach(file) { |line| note << line unless line.match? Regexp.union(IGNORE_LINE) }
  FileUtils.mv(file, TRASH_PATH) && trash_count += 1 if note == template
  FileUtils.mv(file, dir) && archive_count += 1 if File.stat(file).birthtime < Time.now - ARCHIVE_TIME
end

puts "✅ #{archive_count} note(s) archived. #{trash_count} note(s) deleted."